File: Python/Face Recognition/eigenface Recognition using Principal Component Analysis (PCA)/face_recognition.py
import numpy as np
import cv2
import matplotlib.pyplot as plt

loaded = 0  # Initialize the loaded variable
numeric_image = None

def load_database():
    global loaded, numeric_image
    if not loaded:
        all_images = np.zeros((112 * 92, 10 * 10), dtype=np.uint8)
        for i in range(1, 11):
            folder_name = f's{i}'
            for j in range(1, 11):
                image_path = f'{folder_name}/{j}.pgm'
                image_container = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)

                if image_container is None:
                    raise FileNotFoundError(f"Image not found: {image_path}")

                image_container = cv2.resize(image_container, (92, 112))

                all_images[:, (i - 1) * 10 + j - 1] = image_container.flatten()
            print('Loading Database')
        numeric_image = all_images
    loaded = 1
    return numeric_image

# Initialize loaded before calling load_database
loaded_image = load_database()

# Eigenface recognition
image_signature = 20
random_index = int(100 * np.random.rand())
random_image = loaded_image[:, random_index]
rest_of_the_images = np.delete(loaded_image, random_index, axis=1)

white_image = np.ones((1, rest_of_the_images.shape[1]), dtype=np.uint8)
mean_value = np.mean(rest_of_the_images, axis=1, dtype=np.uint8)
mean_removed = rest_of_the_images - np.outer(mean_value, white_image)

L = np.dot(mean_removed.T, mean_removed)
V, D = np.linalg.eig(L)
V = np.dot(mean_removed, V.reshape(-1, 1))  # Reshape V to 2D array
V = V[:, :image_signature]

all_image_signature = np.zeros((rest_of_the_images.shape[1], image_signature), dtype=np.float32)
for i in range(rest_of_the_images.shape[1]):
    all_image_signature[i, :] = np.dot(mean_removed[:, i].T, V)

# Recognition
p = random_image - mean_value
s = np.dot(p.T, V)
z = np.linalg.norm(all_image_signature - s, axis=1, ord=2)
i = np.argmin(z)
accuracy = (1 - z[i] / np.linalg.norm(s, ord=2)) * 100

# Display results (you may need to adjust the display code based on your Python environment)
plt.subplot(121)
plt.imshow(random_image.reshape(112, 92), cmap='gray')
plt.title('Looking for this Face', fontweight='bold', fontsize=16, color='red')

plt.subplot(122)
plt.imshow(rest_of_the_images[:, i].reshape(112, 92), cmap='gray')
plt.title(f'Recognition Completed\nAccuracy: {accuracy:.2f}%', fontweight='bold', fontsize=16, color='red')

plt.show()

File: MATLAB/Face Recognition/eigenface Recognition using Principal Component Analysis (PCA)/face_recognition.m
% Load database
loaded_Image = Load_database();

% Randomly choose an image
random_Index = round(100*rand(1,1));          
random_Image = loaded_Image(:,random_Index);                         
rest_of_the_images = loaded_Image(:,[1:random_Index-1 random_Index+1:end]);         

% Eigenface recognition
image_Signature = 20;                            
white_Image = uint8(ones(1,size(rest_of_the_images,2)));
mean_value = uint8(mean(rest_of_the_images,2));                
mean_Removed = rest_of_the_images-uint8(single(mean_value)*single(white_Image)); 
L = single(mean_Removed)'*single(mean_Removed);
[V,D] = eig(L);
V = single(mean_Removed)*V;
V = V(:,end:-1:end-(image_Signature-1));          
all_image_Signatire = zeros(size(rest_of_the_images,2),image_Signature);
for i = 1:size(rest_of_the_images,2)
    all_image_Signatire(i,:) = single(mean_Removed(:,i))'*V;  
end

% Recognition
p = random_Image - mean_value;
s = single(p)'*V;
z = [];
for i = 1:size(rest_of_the_images,2)
    z = [z,norm(all_image_Signatire(i,:)-s,2)];
end
[~, i] = min(z);
accuracy = (1 - z(i) / (norm(s, 2))) * 100; % calculate accuracy in percentage

% Display results
subplot(121);
imshow(reshape(random_Image,112,92));
title('Looking for this Face','FontWeight','bold','Fontsize',16,'color','red');

subplot(122);
imshow(reshape(rest_of_the_images(:,i),112,92));
title(sprintf('Recognition Completed\nAccuracy: %.2f%%', accuracy), ...
    'FontWeight', 'bold', 'Fontsize', 16, 'color', 'red');
